### PR TITLE
Use stream to dispatch requests for checkouts

### LINKED ISSUE #36: Shut Dow cluenly test fails sometimes
Sometimes this test fails, investigate:

`pool_internal::test::the_pool_shuts_down_cleanly_even_if_connections_cannot_be_created`

### LINKED ISSUE #37: Use stream to dispatch requests for checkouts
Use a stream for the internal checkout synchronization instead of a lock. Also
adjust the configuration and metrics to match the changes.

Get rid of the old configuration values that are not needed anymore since this is almost a rewrite.

### PR DESCRIPTION
**Pull request by chridou**
closes #37 
closes #36 

Internal logic changed completely.

### CODE REVIEW

**Comment by chridou** [comment]
:+1:

File: src/config/builder.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    /// The default is 50.
+    pub fn desired_pool_size(mut self, v: usize) -> Self {
+        self.config.desired_pool_size = v;
+        self
+    }
+
+    /// Sets the behaviour of the pool on checkouts if no specific behaviour
+    /// was requested by the user.
+    pub fn default_checkout_mode<T: Into<DefaultPoolCheckoutMode>>(mut self, v: T) -> Self {
+        self.config.default_checkout_mode = v.into().adjust();
```
Comment (panicbit) [inline src/config/builder.rs]: Remove `.adjust()` etc

File: src/pools/mod.rs [hunk 2]
```diff
@@ -1,6 +1,91 @@
...
+pub(crate) trait CanCheckout<T: Poolable> {
+    /// Directly do a checkout
+    fn check_out<M: Into<CheckoutConstraint>>(&self, constraint: M) -> CheckoutManaged<T>;
+}
+
+/// Retry the checkout if the checkout failed with a
+/// `CheckoutLimitReached` as long as a retry is allowed
+/// by the constraint
+///
+/// TODO: Currently only diverts to a regular checkout
```
Comment (panicbit) [inline src/pools/mod.rs]: Remove the TODO

**Comment by panicbit** [comment]
:+1:

**Pull request merged**